### PR TITLE
rafthttp: version enforcement on rafthttp messages

### PR DESCRIPTION
**Pull request by yichengq**
This PR sets etcd version and min cluster version in request header,
and let server check version compatibility. rafthttp server
will reject any message from peer with incompatible version(too low
version or too high version), and print out warning logs.

for #2716 

The example output for incompatibile etcd members:
```
13:32:28 etcd3 | 2015/06/2 13:32:28 rafthttp: request sent was ignored by peer a8266ecf031671f3 (server version incompatible)
13:32:28 etcd3 | 2015/06/2 13:32:28 rafthttp: roundtripping error: incompatible version
13:32:28 etcd3 | 2015/06/2 13:32:28 rafthttp: request received was ignored (remote version is too low: remote[a8266ecf031671f3]=2.1.0-alpha.0+git, local=2.3.0-alpha.0+git)
13:32:28 etcd3 | 2015/06/2 13:32:28 rafthttp: request received was ignored (remote version is too low: remote[a8266ecf031671f3]=2.1.0-alpha.0+git, local=2.3.0-alpha.0+git)
13:32:28 etcd1 | 2015/06/2 13:32:28 rafthttp: request sent was ignored by peer 924e2e83e93f2560 (server version incompatible)
13:32:28 etcd1 | 2015/06/2 13:32:28 rafthttp: roundtripping error: incompatible version
13:32:28 etcd1 | 2015/06/2 13:32:28 rafthttp: request sent was ignored by peer 924e2e83e93f2560 (server version incompatible)
13:32:28 etcd1 | 2015/06/2 13:32:28 rafthttp: roundtripping error: incompatible version
13:32:28 etcd3 | 2015/06/2 13:32:28 rafthttp: request received was ignored (remote version is too low: remote[6e3bd23ae5f1eae0]=2.1.0-alpha.0+git, local=2.3.0-alpha.0+git)
13:32:28 etcd3 | 2015/06/2 13:32:28 rafthttp: request received was ignored (remote version is too low: remote[6e3bd23ae5f1eae0]=2.1.0-alpha.0+git, local=2.3.0-alpha.0+git)
13:32:28 etcd2 | 2015/06/2 13:32:28 rafthttp: request sent was ignored by peer 924e2e83e93f2560 (server version incompatible)
13:32:28 etcd2 | 2015/06/2 13:32:28 rafthttp: roundtripping error: incompatible version
13:32:28 etcd2 | 2015/06/2 13:32:28 rafthttp: request sent was ignored by peer 924e2e83e93f2560 (server version incompatible)
13:32:28 etcd2 | 2015/06/2 13:32:28 rafthttp: roundtripping error: incompatible version
13:32:28 etcd1 | 2015/06/2 13:32:28 rafthttp: request received was ignored (local version is too low: remote[924e2e83e93f2560]=2.3.0-alpha.0+git, local=2.1.0-alpha.0+git)
13:32:28 etcd1 | 2015/06/2 13:32:28 rafthttp: request received was ignored (local version is too low: remote[924e2e83e93f2560]=2.3.0-alpha.0+git, local=2.1.0-alpha.0+git)
13:32:28 etcd3 | 2015/06/2 13:32:28 rafthttp: request sent was ignored by peer a8266ecf031671f3 (server version incompatible)
13:32:28 etcd3 | 2015/06/2 13:32:28 rafthttp: roundtripping error: incompatible version
13:32:28 etcd3 | 2015/06/2 13:32:28 rafthttp: request sent was ignored by peer a8266ecf031671f3 (server version incompatible)
13:32:28 etcd3 | 2015/06/2 13:32:28 rafthttp: roundtripping error: incompatible version
13:32:28 etcd2 | 2015/06/2 13:32:28 rafthttp: request received was ignored (local version is too low: remote[924e2e83e93f2560]=2.3.0-alpha.0+git, local=2.1.0-alpha.0+git)
13:32:28 etcd2 | 2015/06/2 13:32:28 rafthttp: request received was ignored (local version is too low: remote[924e2e83e93f2560]=2.3.0-alpha.0+git, local=2.1.0-alpha.0+git)
13:32:28 etcd3 | 2015/06/2 13:32:28 rafthttp: request sent was ignored by peer 6e3bd23ae5f1eae0 (server version incompatible)
13:32:28 etcd3 | 2015/06/2 13:32:28 rafthttp: roundtripping error: incompatible version
13:32:28 etcd3 | 2015/06/2 13:32:28 rafthttp: request sent was ignored by peer 6e3bd23ae5f1eae0 (server version incompatible)
13:32:28 etcd3 | 2015/06/2 13:32:28 rafthttp: roundtripping error: incompatible version
13:32:28 etcd3 | 2015/06/2 13:32:28 rafthttp: request received was ignored (remote version is too low: remote[a8266ecf031671f3]=2.1.0-alpha.0+git, local=2.3.0-alpha.0+git)
13:32:28 etcd3 | 2015/06/2 13:32:28 rafthttp: request received was ignored (remote version is too low: remote[a8266ecf031671f3]=2.1.0-alpha.0+git, local=2.3.0-alpha.0+git)
13:32:28 etcd1 | 2015/06/2 13:32:28 rafthttp: request sent was ignored by peer 924e2e83e93f2560 (server version incompatible)
13:32:28 etcd1 | 2015/06/2 13:32:28 rafthttp: roundtripping error: incompatible version
13:32:28 etcd1 | 2015/06/2 13:32:28 rafthttp: request sent was ignored by peer 924e2e83e93f2560 (server version incompatible)
13:32:28 etcd1 | 2015/06/2 13:32:28 rafthttp: roundtripping error: incompatible version
13:32:28 etcd3 | 2015/06/2 13:32:28 rafthttp: request received was ignored (remote version is too low: remote[6e3bd23ae5f1eae0]=2.1.0-alpha.0+git, local=2.3.0-alpha.0+git)
13:32:28 etcd3 | 2015/06/2 13:32:28 rafthttp: request received was ignored (remote version is too low: remote[6e3bd23ae5f1eae0]=2.1.0-alpha.0+git, local=2.3.0-alpha.0+git)
13:32:28 etcd2 | 2015/06/2 13:32:28 rafthttp: request sent was ignored by peer 924e2e83e93f2560 (server version incompatible)
```

TODO:
- [x] add tests

### CODE REVIEW

File: rafthttp/http.go [hunk 1]
```diff
@@ -72,6 +72,18 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if err := checkPeerVersionCompatibility(serverVersion(r.Header), minClusterVersion(r.Header)); err != nil {
+		log.Printf("rafthttp: request ignored due to %v", err)
+		// StatusForbidden has been used to forbid messages from removed member,
```
Comment (xiang90) [inline rafthttp/http.go]: hmm... i think we should still use statusForbidden... the other side can check the string. we need to define a const string error though.

File: rafthttp/http.go [hunk 2]
```diff
@@ -72,6 +72,18 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if err := checkPeerVersionCompatibility(serverVersion(r.Header), minClusterVersion(r.Header)); err != nil {
```
Comment (xiang90) [inline rafthttp/http.go]: checkVerCompability

File: rafthttp/transport.go [hunk 3]
```diff
@@ -33,6 +34,10 @@ type Raft interface {
 	ReportSnapshot(id uint64, status raft.SnapshotStatus)
 }
 
+type ClusterVersion interface {
+	Version() *semver.Version
```
Comment (xiang90) [inline rafthttp/transport.go]: Can we use `ClusterVerison()`?

File: rafthttp/util.go [hunk 4]
```diff
@@ -45,3 +49,73 @@ func readEntryFrom(r io.Reader, ent *raftpb.Entry) error {
...
+	if compareMajorMinorVersion(minCluster, localServer) == 1 {
+		return fmt.Errorf("local version is too low (remote=%s local=%s)", server, localServer)
+	}
+	return nil
+}
+
+// isValidClusterVersionString checks whether the given cluster version string
+// is a valid one in the view of local etcd. It is valid when it is an
+// undecided one, or it is +/-1 of the local version.
+func isValidClusterVersionString(verStr string) bool {
```
Comment (xiang90) [inline rafthttp/util.go]: `checkClusterVersionCompatiblity`?
Can we try to make it consistent with the Version one?

**Comment by xiang90** [comment]
@yichengq We are also spamming the log... We have to have a back-off mechanism for streaming dialing.

**Comment by yichengq** [comment]
all comment are adressed. PTAL

File: rafthttp/http.go [hunk 5]
```diff
@@ -72,6 +74,15 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if err := checkVersionCompability(serverVersion(r.Header), minClusterVersion(r.Header)); err != nil {
+		log.Printf("rafthttp: request ignored due to %v", err)
+		http.Error(w, errMsgIncompatibleVersion, http.StatusForbidden)
+		return
+	}
+	if err := checkClusterVersionCompatiblity(r.Header.Get("X-Cluster-Version")); err != nil {
+		log.Printf("rafthttp: received request that %v", err)
```
Comment (xiang90) [inline rafthttp/http.go]: this log line is very unclear.

File: rafthttp/http.go [hunk 6]
```diff
@@ -72,6 +74,15 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if err := checkVersionCompability(serverVersion(r.Header), minClusterVersion(r.Header)); err != nil {
+		log.Printf("rafthttp: request ignored due to %v", err)
```
Comment (xiang90) [inline rafthttp/http.go]: request ignored (%v)

File: rafthttp/transport.go [hunk 7]
```diff
@@ -33,6 +34,10 @@ type Raft interface {
 	ReportSnapshot(id uint64, status raft.SnapshotStatus)
 }
 
+type ClusterVersion interface {
```
Comment (xiang90) [inline rafthttp/transport.go]: why this is a public interface?

File: rafthttp/pipeline.go [hunk 8]
```diff
@@ -152,6 +166,10 @@ func (p *pipeline) post(data []byte) error {
 			uu.Host, resp.Header.Get("X-Etcd-Cluster-ID"), p.cid)
 		return fmt.Errorf("cluster ID mismatch")
 	case http.StatusForbidden:
+		if string(b) == errMsgIncompatibleVersion {
```
Comment (xiang90) [inline rafthttp/pipeline.go]: Can you use switch instead?

File: rafthttp/http.go [hunk 9]
```diff
@@ -126,6 +137,17 @@ func (h *streamHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 	}
 
+	w.Header().Set("X-Server-Version", version.Version)
+
+	if err := checkVersionCompability(serverVersion(r.Header), minClusterVersion(r.Header)); err != nil {
+		log.Printf("rafthttp: request ignored due to %v", err)
+		http.Error(w, errMsgIncompatibleVersion, http.StatusForbidden)
+		return
+	}
+	if err := checkClusterVersionCompatiblity(r.Header.Get("X-Cluster-Version")); err != nil {
```
Comment (xiang90) [inline rafthttp/http.go]: can you give me an example that this check fails when the check at line 142 passes?

File: rafthttp/http.go [hunk 10]
```diff
@@ -72,6 +74,15 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if err := checkVersionCompability(serverVersion(r.Header), minClusterVersion(r.Header)); err != nil {
+		log.Printf("rafthttp: request ignored due to %v", err)
+		http.Error(w, errMsgIncompatibleVersion, http.StatusForbidden)
```
Comment (xiang90) [inline rafthttp/http.go]: why do we use forbidden in this case? shouldn't this be a 412?

File: rafthttp/http.go (hunk 5 above)
Comment (xiang90) [inline rafthttp/http.go]: can you give me an example that this check fails when the check at line 77 passes?

Comment (yichengq) [inline rafthttp/http.go]: The local version is 2.1, the remote version is 2.2, and the cluster version is 2.0. The local can talk to the remote, but the remote may not be able to talk to some member in the cluster(or he just needs to wait some time for version upgrade).

File: rafthttp/transport.go (hunk 7 above)
Comment (yichengq) [inline rafthttp/transport.go]: It is the type of one argument in NewTransporter function, so it becomes public.

File: rafthttp/pipeline.go (hunk 8 above)
Comment (yichengq) [inline rafthttp/pipeline.go]: will do.

File: rafthttp/http.go (hunk 10 above)
Comment (yichengq) [inline rafthttp/http.go]: I think it fits into both:
403: The server understood the request, but is refusing to fulfill it. Authorization will not help and the request SHOULD NOT be repeated.
412: The precondition given in one or more of the request-header fields evaluated to false when it was tested on the server.

403 sounds more severe and demands users to check version problem. So I choose forbidden here. Any suggestion?

Comment (xiang90) [inline rafthttp/http.go]: This seems exactly what we do: The precondition given in one or more of the request-header fields evaluated to false when it was tested on the server.

File: rafthttp/http.go (hunk 5 above)
Comment (xiang90) [inline rafthttp/http.go]: why a remote 2.2 member can have a cluster version of 2.0?

**Comment by yichengq** [comment]
remove check on cluster-version because it cannot 1. do enforcement on cluster version clearly and easily 2. the warning about cluster version can be done somewhere else.
PTAL

File: rafthttp/http.go [hunk 11]
```diff
@@ -29,6 +29,8 @@ import (
 
 const (
 	ConnReadLimitByte = 64 * 1024
+
+	errMsgIncompatibleVersion = "incompatible version"
```
Comment (xiang90) [inline rafthttp/http.go]: can we define this as error?
so line 163 can return it. and when we want the string we can call err.Error()

File: rafthttp/http.go [hunk 12]
```diff
@@ -34,6 +35,9 @@ const (
 var (
 	RaftPrefix       = "/raft"
 	RaftStreamPrefix = path.Join(RaftPrefix, "stream")
+
+	errIncompatibleVersion = fmt.Errorf("incompatible version")
```
Comment (xiang90) [inline rafthttp/http.go]: errors.New("")

**Comment by yichengq** [comment]
have improved the output format.

**Comment by yichengq** [comment]
changed the line to `request sent was ignored by peer 924e2e83e93f2560 (server version incompatible)`
squashed the commits

**Comment by xiang90** [comment]
LGTM

**Pull request merged**